include/onnxruntime/core/framework/sparse_tensor.h

```diff
@@ -1,72 +1,317 @@
+  SparseRep(MLDataType data_type, const TensorShape& values_shape, void* values, const OrtMemoryInfo& mem_info);
+
+  void* IndicesStart(int64_t align) {
+    return reinterpret_cast<uint8_t*>(values_.MutableDataRaw()) + Roundup(values_.SizeInBytes(), align);
+  }
+
+ public:
+  virtual ~SparseRep();
+
+  ORT_DISALLOW_COPY_AND_ASSIGNMENT(SparseRep);
+
+  const Tensor& Values() const noexcept {
+    return values_;
+  }
+
+  Tensor& MutableValues() noexcept {
+    return values_;
+  }
+
+  bool OwnBuffer() const noexcept {
+    return allocator_ != nullptr;
+  }
+
+  const OrtMemoryInfo& Location() const {
+    return values_.Location();
+  }
+
+  // Round up size to a multiple of align.
+  // Example:
+  //   Roundup(13, 5)   => 15
+  //   Roundup(201, 16) => 208
+  static int64_t Roundup(int64_t size, int64_t align) {
+    return ((size + align - 1) / align) * align;
+  }
+
+  /// <summary>
+  /// Calculate required buffer size. We will place indices
+  /// after data and make sure indices start at int64_t aligned place
+  /// </summary>
+  /// <returns></returns>
+  static int64_t CalculateRequiredBufferSize(int64_t data_size, int64_t indices_size, int64_t align) {
+    return Roundup(data_size, align) + indices_size;
+  }
+
+  /// <summary>
+  ///  Copy the same format to a different destination.
+  /// </summary>
+  /// <param name="data_transfer_manager">manager</param>
+  /// <param name="allocator">allocator for the destination</param>
+  /// <param name="exec_q_id"></param>
+  /// <param name="dst_rep">[out] destination representation</param>
+  /// <returns>status</returns>
+  virtual Status Copy(const DataTransferManager& data_transfer_manager, const AllocatorPtr& allocator,
+                      int exec_q_id, std::unique_ptr<SparseRep>& dst_rep) const = 0;
+
+  virtual Status Copy(const IDataTransfer& data_transfer, const AllocatorPtr& allocator, int exec_q_id,
+                      std::unique_ptr<SparseRep>& dst_rep) const = 0;
+
+  /// <summary>
+  /// Calculates and returns how much this fully initialized SparseTensor data (would)
+  /// occupy in a contiguous allocation block
+  /// </summary>
+  /// <returns></returns>
+  virtual int64_t RequiredAllocationSize() const noexcept = 0;
+
+ private:
+  void ReleaseBuffer();
+
+  AllocatorPtr allocator_;  // Used to allocate p_data_ or nullptr
+  void* p_data_;
+  Tensor values_;
+};
+
 /**
  * @brief This class implements SparseTensor. 
- * We represent a SparseTensor as a triple <values, indices, shape>. "values" and "indices" themselves
+ * We represent a SparseTensor COO as a triple <values, indices, shape>. "values" and "indices" themselves
  * are implemented as Tensors. 
  * We follow the Tensor design for memory ownership/management: a sparse-tensor does not own the "value"
  * or "indices" tensors.
  */
 
 class SparseTensor final {
+  SparseTensor(MLDataType elt_type,
+               const TensorShape& dense_shape,
+               const OrtMemoryInfo& location,
```

Comment:
Why do we need to pass OrtMemoryInfo if we've the allocator in the interface?

Reply:
This is a private ctor. Sometimes we pass only location, other times we pass allocator, and we populate location from allocator. It is similar to Tensor.